File: test/java/brs/web/api/http/handler/GetAllAssetsTest.java
package brs.web.api.http.handler;

import brs.Asset;
import brs.Blockchain;
import brs.Signum;
import brs.assetexchange.AssetExchange;
import brs.common.AbstractUnitTest;
import brs.common.QuickMocker;
import brs.common.QuickMocker.MockParam;
import brs.services.AccountService;
import brs.util.CollectionWithIndex;
import brs.util.JSON;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import javax.servlet.http.HttpServletRequest;
import java.util.Collection;

import static brs.web.api.http.common.Parameters.FIRST_INDEX_PARAMETER;
import static brs.web.api.http.common.Parameters.LAST_INDEX_PARAMETER;
import static brs.web.api.http.common.ResultFields.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.*;

@RunWith(PowerMockRunner.class)
@PrepareForTest(Signum.class)
public class GetAllAssetsTest extends AbstractUnitTest {

  private GetAllAssets t;

  private AssetExchange assetExchange;
  private AccountService accountService;

  @Before
  public void setUp() {
    assetExchange = mock(AssetExchange.class);
    accountService = mock(AccountService.class);

    mockStatic(Signum.class);
    Blockchain mockBlockchain = mock(Blockchain.class);
    when(Signum.getBlockchain()).thenReturn(mockBlockchain);
    when(mockBlockchain.getHeight()).thenReturn(Integer.MAX_VALUE);

    t = new GetAllAssets(assetExchange, accountService);
  }

  @Test
  public void processRequest() {
    final int firstIndex = 1;
    final int lastIndex = 2;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(FIRST_INDEX_PARAMETER, firstIndex),
        new MockParam(LAST_INDEX_PARAMETER, lastIndex)
    );

    final long mockAssetId = 1;

    final Asset mockAsset = mock(Asset.class);
    when(mockAsset.getId()).thenReturn(1L);
    when(mockAsset.getId()).thenReturn(mockAssetId);
    when(mockAsset.getName()).thenReturn("name");
    when(mockAsset.getDescription()).thenReturn("description");
    when(mockAsset.getDecimals()).thenReturn((byte) 1);
    when(mockAsset.getQuantityQnt()).thenReturn(2L);

    final Collection<Asset> mockAssetIterator = mockCollection(mockAsset);

    when(assetExchange.getAllAssets(eq(firstIndex), eq(lastIndex))).thenReturn(new CollectionWithIndex<Asset>(mockAssetIterator, -1));
    when(assetExchange.getAssetAccountsCount(eq(mockAsset), eq(0L), eq(true), eq(true))).thenReturn(1);
    when(assetExchange.getTransferCount(eq(mockAssetId))).thenReturn(2);
    when(assetExchange.getTradeCount(eq(mockAssetId))).thenReturn(3);

    final JsonObject result = (JsonObject) t.processRequest(req);
    assertNotNull(result);

    final JsonArray assetsResult = (JsonArray) result.get(ASSETS_RESPONSE);
    assertNotNull(assetsResult);
    assertEquals(1, assetsResult.size());

    final JsonObject assetResult = (JsonObject) assetsResult.get(0);
    assertNotNull(assetResult);

    assertEquals(mockAsset.getName(), JSON.getAsString(assetResult.get(NAME_RESPONSE)));
    assertEquals(mockAsset.getDescription(), JSON.getAsString(assetResult.get(DESCRIPTION_RESPONSE)));
    assertEquals(mockAsset.getDecimals(), JSON.getAsByte(assetResult.get(DECIMALS_RESPONSE)));
    assertEquals("" + mockAsset.getQuantityQnt(), JSON.getAsString(assetResult.get(QUANTITY_QNT_RESPONSE)));
    assertEquals("" + mockAsset.getId(), JSON.getAsString(assetResult.get(ASSET_RESPONSE)));
  }

}

File: test/java/brs/http/GetAllAssetsTest.kt
package brs.http

import brs.Asset
import brs.assetexchange.AssetExchange
import brs.common.AbstractUnitTest
import brs.common.QuickMocker
import brs.common.QuickMocker.MockParam
import brs.util.JSON
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import org.junit.Before
import org.junit.Test

import javax.servlet.http.HttpServletRequest

import brs.http.common.Parameters.FIRST_INDEX_PARAMETER
import brs.http.common.Parameters.LAST_INDEX_PARAMETER
import brs.http.common.ResultFields.*
import com.nhaarman.mockitokotlin2.eq
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever

class GetAllAssetsTest : AbstractUnitTest() {

    private var t: GetAllAssets? = null

    private var assetExchange: AssetExchange? = null

    @Before
    fun setUp() {
        assetExchange = mock<AssetExchange>()

        t = GetAllAssets(assetExchange!!)
    }

    @Test
    fun processRequest() {
        val firstIndex = 1
        val lastIndex = 2

        val req = QuickMocker.httpServletRequest(
                MockParam(FIRST_INDEX_PARAMETER, firstIndex),
                MockParam(LAST_INDEX_PARAMETER, lastIndex)
        )

        val mockAssetId: Long = 1

        val mockAsset = mock<Asset>()
        whenever(mockAsset.id).thenReturn(1L)
        whenever(mockAsset.id).thenReturn(mockAssetId)
        whenever(mockAsset.name).thenReturn("name")
        whenever(mockAsset.description).thenReturn("description")
        whenever(mockAsset.decimals).thenReturn(1.toByte())
        whenever(mockAsset.quantityQNT).thenReturn(2L)

        val mockAssetIterator = mockCollection<Asset>(mockAsset)

        whenever(assetExchange!!.getAllAssets(eq(firstIndex), eq(lastIndex))).thenReturn(mockAssetIterator)
        whenever(assetExchange!!.getAssetAccountsCount(eq(mockAssetId))).thenReturn(1)
        whenever(assetExchange!!.getTransferCount(eq(mockAssetId))).thenReturn(2)
        whenever(assetExchange!!.getTradeCount(eq(mockAssetId))).thenReturn(3)

        val result = t!!.processRequest(req) as JsonObject
        assertNotNull(result)

        val assetsResult = result.get(ASSETS_RESPONSE) as JsonArray
        assertNotNull(assetsResult)
        assertEquals(1, assetsResult.size().toLong())

        val assetResult = assetsResult.get(0) as JsonObject
        assertNotNull(assetResult)

        assertEquals(mockAsset.name, JSON.getAsString(assetResult.get(NAME_RESPONSE)))
        assertEquals(mockAsset.description, JSON.getAsString(assetResult.get(DESCRIPTION_RESPONSE)))
        assertEquals(mockAsset.decimals.toLong(), JSON.getAsByte(assetResult.get(DECIMALS_RESPONSE)).toLong())
        assertEquals("" + mockAsset.quantityQNT, JSON.getAsString(assetResult.get(QUANTITY_QNT_RESPONSE)))
        assertEquals("" + mockAsset.id, JSON.getAsString(assetResult.get(ASSET_RESPONSE)))
        assertEquals(1, JSON.getAsInt(assetResult.get(NUMBER_OF_ACCOUNTS_RESPONSE)).toLong())
        assertEquals(2, JSON.getAsInt(assetResult.get(NUMBER_OF_TRANSFERS_RESPONSE)).toLong())
        assertEquals(3, JSON.getAsInt(assetResult.get(NUMBER_OF_TRADES_RESPONSE)).toLong())
    }

}
